### DISCUSSION TITLE
Plugin import and registration.

### DISCUSSION THREAD

**Discussion by dbauszus-glx** [chartjs/chartjs-plugin-datalabels#290] [Q&A] [answered]
I have npm installed chart.js (^3.5.1) and chartjs-plugin-datalabels (^2.0.0).

I got stuck at trying to a dynamic module import. https://stackoverflow.com/questions/69543515/dynamic-import-of-chart-js-with-plugins

Having failed to do a dynamic import I tested the regular import from local install but this doesn't work neither.

The [integration notes for this module](https://chartjs-plugin-datalabels.netlify.app/guide/getting-started.html#integration) use following example.

```
import {Chart} from 'chart.js';
import ChartDataLabels from 'chartjs-plugin-datalabels';
```

I don't think this can work.

I can use the auto registration import for chart.js to use the core library.

```
import Chart from 'chart.js/auto';
```

Importing and registering the plugin doesn't seem to be possible though.

```
import Chart from 'chart.js/auto';

import ChartDataLabels from 'chartjs-plugin-datalabels';

Chart.register(ChartDataLabels)
```

This will fail because of the chartjs helpers.

Any advice on the import of this and indeed any chartjs module?

**Comment by dbauszus-glx** [comment]
This can be closed. Import and registration do work. Just not for pie charts. Odd. I will open a bug report with chart.js the main repo.

**Comment by dbauszus-glx** [comment]
The documentation for this plugin is still wrong though and should use import from chart.js/auto

**Comment by dbauszus-glx** [comment]
Actually I am not so sure whether this is a bug in chart-js main repo or in this plugin.

The plugin registration is trying to read something from the x-axis which doesn't exist in a pie chart.

I updated my codepen. This works with a bar chart. Change the chart type to pie and the plugin registration breaks with a debug warning in regards to the x-scale. If however added from script tag the plugin works on a pie chart. 🤷

Any advise?

**Comment by simonbrunel** [comment, marked as answer]
Somehow, there is multiple versions of `chart.js` imported in your Codepen:

```js
await import("https://cdn.skypack.dev/chart.js") 
// imports chart.js@v3.5.1
```
```js
await import("https://cdn.skypack.dev/chartjs-plugin-datalabels") 
// imports chartjs-plugin-datalabels@v2.0.0
    // which imports chart.js@3.5.0
```

That means the plugin uses classes from a different version of `chart.js` than the one used to create the chart. And this is why it [fails to determine the type of element](https://github.com/chartjs/chartjs-plugin-datalabels/blob/master/src/label.js#L59) and uses the `fallback` positioner, which doesn't work with `ArcElement`. 

It's not a bug in this lib but an issue with the way imports are resolved. I don't know why skypack picks chart.js 3.5.0 instead of 3.5.1 when importing the plugin but maybe there is a way to configure it. As a workaround (and probably a good practice), you can import specific version of both packages:

```js
const { Chart, registerables } = await import("https://cdn.skypack.dev/chart.js@3.5.0");
const plugin = await import("https://cdn.skypack.dev/chartjs-plugin-datalabels@2.0.0");

Chart.register(...registerables, plugin.default);
```
https://codepen.io/simonbrunel/pen/PoKqZBQ

Though, I'm not sure that `chartjs-plugin-datalabels@2.0.0` will always import `chart.js@3.5.0` using this skypack link.

**Reply by dbauszus-glx** [reply]
Excellent find. Let me see whether skypack can rebuild the plugin.

I assume the build is triggered on release.

Chart.js 3.5.1 was released on 18th August. https://www.skypack.dev/view/chart.js

Plugin-Datalabels 2.0.0 was released on 1st July https://www.skypack.dev/view/chartjs-plugin-datalabels

This would explain why the skypack build process would have used an older version of the library.

**Reply by dbauszus-glx** [reply]
It's an issue with skypack pinning dependencies. As outlined in this issue. https://github.com/skypackjs/skypack-cdn/issues/86

I marked your answer as this put us on the right track. Nothing we can do here. It's down to the skypack team to sort this out.
